libs/common/src/models/domain/encArrayBuffer.ts

```
@@ -1,3 +1,53 @@
+import { EncryptionType } from "@bitwarden/common/enums/encryptionType";
+
 export class EncArrayBuffer {
-  constructor(public buffer: ArrayBuffer) {}
+  readonly encryptionType: EncryptionType = null;
+
+  readonly ctBytes: ArrayBuffer = null;
+  readonly ivBytes: ArrayBuffer = null;
+  readonly macBytes: ArrayBuffer = null;
+
+  constructor(readonly buffer: ArrayBuffer) {
+    const encBytes = new Uint8Array(buffer);
+    const encType = encBytes[0];
+
+    switch (encType) {
+      case EncryptionType.AesCbc128_HmacSha256_B64:
+      case EncryptionType.AesCbc256_HmacSha256_B64:
+        if (encBytes.length <= 49) {
+          // 1 + 16 + 32 + ctLength
+          return;
+        }
+
+        this.encryptionType = encType;
+        this.ivBytes = encBytes.slice(1, 17).buffer;
+        this.macBytes = encBytes.slice(17, 49).buffer;
+        this.ctBytes = encBytes.slice(49).buffer;
+        break;
+      case EncryptionType.AesCbc256_B64:
+        if (encBytes.length <= 17) {
+          // 1 + 16 + ctLength
+          return;
+        }
+
+        this.encryptionType = encType;
+        this.ivBytes = encBytes.slice(1, 17).buffer;
+        this.ctBytes = encBytes.slice(17).buffer;
+        break;
+      default:
+        return;
```

It is preferrable to have types throw that cannot initialize rather than silently not populate.

The silent model produces null errors down the line that are difficult to trace. Loud errors tell you exactly where the problem is happening.

---

Good point about not propagating nulls. I was focusing on replicating current behaviour, but this shouldn't cause any issues, it should just enforce current assumptions.

What do you think about parsing in a private method instead, which is called by getters? e.g.

```ts
private parse(): { iv: ArrayBuffer, mac: ArrayBuffer, data: ArrayBuffer } {
  // current parsing logic here, returns the object described above
  // this keeps the parsing logic all in one place, and array slicing is cheap
}

get macBytes() {
  return this.parse().mac;
}

get ivBytes() {
 return this.parse().iv;
}
```

We could still check for a valid input in the constructor and throw if required.

I'm also not 100% sure that the current logic won't create duplicate arrays in memory. Depends on Javascript memory management. But I think this would be a bit safer in that respect as well.

---

We spoke on the phone about this, but the call to make is this: Should this class exist if it isn't parseable (and therefore not decryptable)? There's an argument either way. I'm OK with either.

---

I'm glad you mentioned this, because it made me realise that I've broken existing behaviour.

Currently, if the `ArrayBuffer` can't be parsed, `cryptoService.decryptFromBytes` returns `null`, and the caller checks for this.

Now that it's moved to the constructor, the ctor will always return a not-null value, so these null checks will no longer work.

I've changed this to throw explicitly. `cryptoService.encryptToBytes` wraps it in a `try/catch` and returns `null` to preserve existing behaviour.

I also prefer the explicit throw generally, we overuse `null` for this at the moment generally.

---

I also added a test for `CryptoService` to make sure this works, and as a form of documenting this behaviour.